### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by ernestmusong**
In this PR, we have created the following pages:

A login page.
A registration page.
A food list (with all CRUD implementation, except for 'update).
A recipes list (with all CRUD implementation, except for 'update').
A general shopping list view (all missing food for all your recipes and total price).
A list of all public recipes from other users with their names and total prices.
Write tests using RSpec and Capybara testing libraries

### CODE REVIEW

File: app/controllers/foods_controller.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  private
+
+  def set_food
+    @food = Food.find(params[:id])
+  end
+
+  def food_params
+    params.require(:food).permit(:name, :measurement_unit, :price, :quantity)
+  end
+end
```
Comment (muneebulrehman) [inline app/controllers/foods_controller.rb]: Your food_controller is not behaving as it should. I have created [this ](https://www.loom.com/share/c15d089a799e485bab9efb56c61429dd?sid=15602e16-d36e-46d1-a72b-5eac8750963a)video where i explained what the issue is. Any food item that any user adds is accessible to all the users and any user can delete it. That should not be the case.

File: app/controllers/recipe_foods_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  private
+
+  def set_recipe_food
+    @recipe_food = RecipeFood.includes(:recipe, :food).find(params[:id])
+  end
+
+  def recipe_food_params
+    params.require(:recipe_food).permit(:quantity, :recipe_id, :food_id)
+  end
+end
```
Comment (muneebulrehman) [inline app/controllers/recipe_foods_controller.rb:64]: When creating a recipe, I am not able to add the food from the foodlist. Instead it creats a new food item which then is visible in all the recipes. And If I have some food items in my food list, they automatically show as food items under the recipe. I have created [this ](https://www.loom.com/share/c15d089a799e485bab9efb56c61429dd?sid=15602e16-d36e-46d1-a72b-5eac8750963a)video where i explained what the issue is

File: app/controllers/shopping_lists_controller.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class ShoppingListsController < ApplicationController
+  def index
+    @users = User.all.includes(:foods)
+    @user = User.includes(:foods).find(params[:user_id])
+    @foods = Food.where.not(user_id: current_user.id).includes(:user)
+  end
+end
```
Comment (muneebulrehman) [inline app/controllers/shopping_lists_controller.rb:7]: Your shopping list is not displaying the correct data. I have recorded [this ](https://www.loom.com/share/c15d089a799e485bab9efb56c61429dd?sid=15602e16-d36e-46d1-a72b-5eac8750963a)video where i explained what the issue is

**Review by muneebulrehman — CHANGES_REQUESTED** [review]
Hi team, 

Good job so far! I see you put a lot of effort into this project. However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- No linter errors
- Descriptive PR
- Good docs

## Required Changes ♻️
_Check the comments under the review._

- [ ] A lot of your tests are failing. Kindly consider fixing these tests before submitting for re-review. If all the tests are passing on your side, then you can submit it without any changes and the next reviewer will verify it. However, I would request you to run the tests twice on your and your partner's machine, just to be sure.
image


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/public_recipes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class PublicRecipesController < ApplicationController
+  def index
+    @recipes = Recipe.where(public: true).order(created_at: :desc).includes(:user)
+  end
```
Comment (muneebulrehman) [inline app/controllers/public_recipes_controller.rb:4]: If a recipe is marked as public by the owner, any other user can mark it back as private. Any user is also able to add food items to a public recipe. That should not be the case. Kindly fix this issue.

File: app/views/foods/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  <td><%= food.measurement_unit%></td>
+  <td>$<%= food.price%></td>
+  <td>
+    <%= link_to "Delete", user_food_path(user_id: food.user.id, id: food.id), class: "delete_btn", method: :delete %>
+  </td>
+</tr>
+    <% end %>
+    </table>
+    <% end %>
+</main>
```
Comment (muneebulrehman) [inline app/views/foods/index.html.erb]: In the food index page, there should be a button that should lead to the form that allows users to add new food. But this button is missing in your case. Kindly fix this issue. 
image

File: app/views/public_recipes/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+            <div class="item">
+                <p>total food items: <strong><%= recipe.user.foods.count %></strong></p>
+                <p>total price: <strong>$<%= recipe.user.foods.sum { |food| food.price * food.quantity } %></strong></p>
+            </div>
+        </li>
+      <%end%>
+    <% end %>
+    </ul>
+    <% end %>
+</main>
```
Comment (muneebulrehman) [inline app/views/public_recipes/index.html.erb:23]: When I am not logged in, I am still able to see the public recipes, and when I click on the recipe, it leads to this error.
image

Comment (ernestmusong) [reply app/views/public_recipes/index.html.erb:23]: Thank you @muneebulrehman for pointing out this error. However, the error has been fixed but I have deliberately allowed the public recipes to be accessible by all users both logged-in and non logged-in users. In this case, non logged-in users have read-only capabilities.

File: app/controllers/recipe_foods_controller.rb (hunk 2 above)
Comment (ernestmusong) [reply app/controllers/recipe_foods_controller.rb:64]: > When creating a recipe, I am not able to add the food from the foodlist. Instead it creats a new food item which then is visible in all the recipes. And If I have some food items in my food list, they automatically show as food items under the recipe. I have created [this ](https://www.loom.com/share/c15d089a799e485bab9efb56c61429dd?sid=15602e16-d36e-46d1-a72b-5eac8750963a)video where i explained what the issue is

[Based on this requirement](https://github.com/microverseinc/curriculum-rails/blob/main/recipe-app/buisness_requirements.md#:~:text=If%20the%20user%20is%20the%20owner%20of%20the%20recipe%2C%20should%20lead%20to%20the%20form%20that%20allows%20the%20user%20to%20add%20new%20food), I think the button should lead to the form that allows the user to add new food. This new food is added to the users food list.

**Comment by ernestmusong** [comment]
Thank you for your overall feedback, the above requirements have been implemented.

File: app/controllers/foods_controller.rb (hunk 1 above)
Comment (rloterh) [reply app/controllers/foods_controller.rb]: well resolved! ✔️

File: app/controllers/shopping_lists_controller.rb (hunk 3 above)
Comment (rloterh) [reply app/controllers/shopping_lists_controller.rb:7]: well resolved! ✔️

File: app/controllers/public_recipes_controller.rb (hunk 4 above)
Comment (rloterh) [reply app/controllers/public_recipes_controller.rb:4]: well resolved! ✔️

File: app/views/public_recipes/index.html.erb (hunk 6 above)
Comment (rloterh) [reply app/views/public_recipes/index.html.erb:23]: well resolved! ✔️

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Musong & Brhan,  

Your project is complete! 🏁 Nice job! There is nothing else to say other than... it's time to merge it 🍾 Congratulations!  🎉💯🌟 

## Highlights 🎯 
🎲 Good job on implementing recommended changes!
🎲 Recipe app is working as intended

Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**